Implement dartPluginClass support for plugins

blasten opened the pull request:
Tool side of https://github.com/flutter/flutter/issues/52267

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test exempt.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

Commit:
Prototype code in the tool

blasten commented:
@jonahwilliams what do you think of this approach?  It supports hot-reload, etc...  If it makes sense, I will still need to get the metadata from pubspec.yaml, which I don't think it's accessible at this point.

Commit:
clean up

jonahwilliams commented on packages/flutter_tools/lib/src/compile.dart:
```diff
@@ -300,6 +311,27 @@ class KernelCompiler {
+  print('_registerPlugins is called');
+}
+
+void main() {
```
If you're going to generate a synthetic entrypoint, then I don't think you need the pragma or anything else special - just invoke _registerPlugins() before calling entrypoint.main.

blasten replied:
main can be overridden by the embedding at runtime.

jonahwilliams replied:
right :) . lots of moving parts here.

jonahwilliams commented:
IIRC this won't work as is without more changes in the resident_runner or the hot reload pipeline. The development compiler needs to be given the same entrypoint to compile as the app was built with - otherwise I don't believe hot reloads will work until a hot restart - and in this case that hot restart would blow away the plugin callback.

This will also require some special casing for g3, since they'll likely generate that file in a different way.

Commit:
Handle hot restart

Commit:
Resolve plugin interfaces

Commit:
Wrong comment

Commit:
Format

Commit:
final

Commit:
Add throwOnPluginPubspecError flag

Commit:
Make isDirectDependency optional

Commit:
Check if dependencies is null

Commit:
Merge branch 'master' into tool_dart_plugin_registrant

Commit:
Provide default package

Commit:
null check

Commit:
Check pubspec.yaml exist

Commit:
Refactor

Commit:
Fix tests

Commit:
Fix assert

Commit:
Compilers reuse same entrypoint

jonahwilliams commented on packages/flutter_tools/lib/src/compile.dart:
```diff
@@ -17,6 +17,8 @@ import 'base/logger.dart';
 import 'base/platform.dart';
 import 'build_info.dart';
 import 'convert.dart';
+import 'globals.dart' as globals;
```
please don't introduce more globals usage

jonahwilliams replied:
either inject a helper interface to do the plugin implementation, or provide the globally used interfaces directly

blasten replied:
Done

jonahwilliams commented on packages/flutter_tools/test/general.shard/build_system/targets/common_test.dart:
```diff
@@ -71,7 +71,7 @@ void main() {
       throwsA(isA<MissingDefineException>()));
   });
 
-  testWithoutContext('KernelSnapshot handles null result from kernel compilation', () async {
```
please change these back

blasten replied:
Done

jonahwilliams reviewed (commented):
Still looking at actual implementation, first comment is not to introduce more globals usage

jonahwilliams reviewed (commented):
General feedback: it is difficult to do a thorough review without tests, as I'm not quite sure what all of the expected changes are

Commit:
Remove global and context usage

Commit:
Tests

jonahwilliams commented on packages/flutter_tools/lib/src/compile.dart:
```diff
@@ -577,7 +599,16 @@ class DefaultResidentCompiler implements ResidentCompiler {
-
+    // Write `generated_main.dart` under the tool's owned directory.
+    // This ensures that the `lib` directory isn't polluted with generated files.
+    final Directory buildDir = _fileSystem.file(mainUri.path)
```
The mainUri could be anywhere, so this isn't stable. I think the correct generated directory should be available on the FlutterProject

blasten replied:
Done

jonahwilliams commented on packages/flutter_tools/lib/src/compile.dart:
```diff
@@ -577,7 +599,16 @@ class DefaultResidentCompiler implements ResidentCompiler {
+        .parent
+        .childDirectory(_fileSystem.path.join('.dart_tool', 'flutter_build'));
+    final File newMainDart = buildDir.childFile('generated_main.dart');
+    if (newMainDart.existsSync()) {
```
In what cases does this not exist?

blasten replied:
If there aren't plugins or if no plugin needs a Dart registry

jonahwilliams commented on packages/flutter_tools/lib/src/plugins.dart:
```diff
@@ -306,14 +342,21 @@ class Plugin {
   final String path;
 
+  /// The name of the interface package that this plugin implements.
+  final String implement;
```
Consider a more descriptive name. "implementsPackage"? "packageImplements"? not quite sure.

If this doesn't implement a package, is it null?

blasten replied:
Done

jonahwilliams commented on packages/flutter_tools/lib/src/plugins.dart:
```diff
@@ -351,21 +395,165 @@ Future<List<Plugin>> findPlugins(FlutterProject project, { bool throwOnError = t
     project.directory.path,
     '.packages',
   );
+  final String pubspec = globals.fs.path.join(
```
This sort of parsing should be a part of the FlutterProject or FlutterManifest classes, since those already have the mechanisms for validating the yaml.

blasten replied:
It validates the Flutter plugins pubspec in conjunction with the app's pubspec.

jonahwilliams replied:
I'm not quite sure what you mean.

jonahwilliams replied:
Please don't re-parse the pubspec here. Lookup this information from the FlutterProject and add it if it is missing.

jonahwilliams commented on packages/flutter_tools/lib/src/plugins.dart:
```diff
@@ -351,21 +395,165 @@ Future<List<Plugin>> findPlugins(FlutterProject project, { bool throwOnError = t
+      if (plugin.implement == null || plugin.implement.isEmpty) {
+        final String defaultImplementation = platformEntries[kDefaultPackage] as String;
+        if (defaultImplementation == null) {
+          if (throwOnPluginPubspecError) {
```
In the event that the user as N plugins that all have this same problem, immediately throwing the error will force them to resolve them one at a time. The pattern used in the asset manifest and pubspec validation is to printError and continue to the end, throwing if you've encountered at least one error.

blasten replied:
Done

jonahwilliams commented on packages/flutter_tools/lib/src/plugins.dart:
```diff
@@ -351,21 +395,165 @@ Future<List<Plugin>> findPlugins(FlutterProject project, { bool throwOnError = t
+///
+///  For more details, https://flutter.dev/go/federated-plugins.
+List<PluginInterfaceResolution> resolvePlatformImplementation(
+  List<Plugin> plugins, {
```
I don't recall if plugins are sorted by this point, but since the order matters for the errors when their are conflicts, you should ensure they are sorted so the error messages stay consistent.

blasten replied:
The order of the plugins doesn't affect the entries in the output list. The order of the entries would differ, but that isn't a problem.

jonahwilliams commented on packages/flutter_tools/lib/src/plugins.dart:
```diff
@@ -797,6 +1033,45 @@ void registerPlugins(Registrar registrar) {
 }
 ''';
 
+// TODO(egarciad): Evaluate merging the web and desktop plugin registry templates.
```
This needs a language version header. Because it isn't under `lib/` it isn't a part of any package and will take whatever the current default language version is. That may be incorrect for users that have not yet migrated to null safety.

In this case, you should pick whatever the language version of the main entrypoint is.

blasten replied:
good point

blasten replied:
Done. It now sets the language version returned from `determineLanguageVersion`.

Commit:
Fix check

Commit:
Fix test

jonahwilliams commented on packages/flutter_tools/lib/src/plugins.dart:
```diff
@@ -797,6 +1033,45 @@ void registerPlugins(Registrar registrar) {
+// Generated file. Do not edit.
+//
+
+import '{{mainEntrypoint}}' as entrypoint;
```
What happens if the user adds a new plugin and does a hot reload/hot restart?
What happens if the user removes a plugin and does a hot restart?

In both cases the file needs to be regenerated, in the former its definitely a nice to have, but in a later that would require a quit/restart cycle to remove a plugin - because the compiler can't resolve it anymore.

What happens if the plugin declaration is invalid in a way that doesn't compile, what do those error messages look like?

jonahwilliams replied:
But in this case, if its a dart plugin and you remove it - you need to regenerate this file or else the app won't compile at all.

blasten replied:
so the case is:  as a user I have a Dart only plugin, run `flutter run`, then delete the dart plugin from pubspec.yaml, then hot restart?

So the workaround is to re-run `flutter run`.  I tried to regenerate the file in the recompile flow, but it required more refactoring.

jonahwilliams replied:
I think there could be more complicated scenarios too - like you update a version and that goes from being a Dart only plugin to not. Unfortunately the error message is going to be very difficult to understand, since the user is going to get a compilation error pointing them to code they didn't write

blasten replied:
yeah. I didn't know that `flutter run` supports adding/removing dependencies to pubpec.yaml if they don't need platform-side recompilation.  I also wonder if users are able to know the difference. e.g. If they consume a plugin, are they able to know that they can't do that for Android and iOS, but it might be ok for a Dart-only plugin.

jonahwilliams replied:
I think removing a plugin needs to be handled otherwise it will become a compilation error.

blasten replied:
Removing a plugin works, they would just need to type `flutter run` again.

jonahwilliams replied:
while not being able to add a new plugin is in-line with our current behavior on Android/iOS, losing the ability to remove a plugin without restarting the tool is definitely a regression.

blasten replied:
I started by handling this case (support adding and removing plugins). Unfortunately, it required additional changes to the test. See your comment about not introducing more globals or using `testUsingContext`.  My suggestion is two separate this feature into two PRs.

Commit:
Fixes

Commit:
Fix import

jonahwilliams commented:
Because this change affects the hot reload/restart development cycle, its critical that we also get some integration tests running it with actual dart plugins wired up - perhaps synthetic to reduce dependencies.

This could either be done in the devicelab hostonly, or you could add an affordance for wiring up dart plugins for the flutter tester device and run in the tools integration shard

blasten commented:
@jonahwilliams are there any `flutter run` e2e test for desktop?

jonahwilliams commented:
I believe the standard hot reload/hot restart tests run on all platforms now (@stuartmorgan ?) but they do not have plugins.

stuartmorgan commented:
> I believe the standard hot reload/hot restart tests run on all platforms now (@stuartmorgan ?) but they do not have plugins.

It runs for macOS, I'm very close on Windows, Linux is currently blocked on some infra help getting the headless VMs to be able to launch Flutter apps. (The test case I'm working on is the devicelab hot reload test that you brought up for macOS a while back; currently we don't have any devicelab tests for Windows and Linux).

stuartmorgan commented:
We could get it running for macOS now though, and then bring Windows and Linux up soon once the desktop devicelab infrastructure is in place, assuming it's mostly cross-platform code.

Commit:
More fixes

Commit:
Remove fileSystem arg

Commit:
projectRootPath can be null

jonahwilliams commented:
Something else that needs to be considered - suppose a plugin incorrectly implements this sort of registerWith functionality (typo or otherwise). Are we just going to spit out a compilation error (in a file most users will not know about)? Do we have any publishing validation that these interfaces are implemented? Maybe we could do more runtime checks in a "debug" mode so that these errors could be handled?

jonahwilliams commented:
Also, if the engine is going to invoke this method - where do the errors go?

blasten commented:
> Something else that needs to be considered - suppose a plugin incorrectly implements this sort of registerWith functionality (typo or otherwise)

@jonahwilliams how is it different than Android & iOS in regard to consuming a 3P dependency?

jonahwilliams commented:
Today if someone adds a dependency on a bad plugin for iOS/android what happens?

blasten commented:
If there's an error in the static `registerWith` method, it shows the stacktrace that includes the generated registry file.

In the case of a Dart registry, this is what the error looks like:
```
E/flutter ( 9432): #0      _registerPlugins (.../.dart_tool/flutter_build/generated_main.dart:14:3)
E/flutter ( 9432): #1      _runMainZoned.<anonymous closure>.<anonymous closure> (dart:ui/hooks.dart:135:29)
E/flutter ( 9432): #2      _rootRun (dart:async/zone.dart:1354:13)
E/flutter ( 9432): #3      _CustomZone.run (dart:async/zone.dart:1258:19)
E/flutter ( 9432): #4      _runZoned (dart:async/zone.dart:1788:10)
E/flutter ( 9432): #5      runZonedGuarded (dart:async/zone.dart:1776:12)
E/flutter ( 9432): #6      _runMainZoned.<anonymous closure> (dart:ui/hooks.dart:133:5)
E/flutter ( 9432): #7      _delayEntrypointInvocation.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:283:19)
E/flutter ( 9432): #8      _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:184:12)
```

jonahwilliams commented:
Have you thought about how this is going to work with existing error reporting solutions like crashlytics or other Zone based error handlers?

blasten commented:
I'd expect errors of this kind to be very uncommon, and only possible if a plugin author didn't bother testing the example app. Even more unlikely would be to see such plugin gaining usage.

jonahwilliams commented:
to handle the case of static errors, in debug mode you could wrap the whole thing in a try catch that caught a NSM, and then used some additional info to point users to offending plugins.

blasten commented:
Somehow:
```
try {
  /// something
 } on NoSuchMethodError catch(error) {
    print('Dart plugins had an error: \${error}');
}
```

didn't work. Is this correct?

jonahwilliams commented:
For debug mode you could catch everything instead, but this should be easy to simulate locally to get something working (i.e. in dart pad)

jonahwilliams commented:
Oh right, it will be a compile error if the method is missing/misspelled

jonahwilliams commented:
so there isn't anything to catch

Commit:
Tool test gets confused

Commit:
Fix test

Commit:
Add e2e test

Commit:
Format a bit nicer

blasten commented:
I added the e2e test. The test needs to be checked in prior to enabling it on LUCI. PTAL

blasten commented on packages/flutter_tools/lib/src/build_system/targets/common.dart:
```diff
@@ -273,6 +273,8 @@ class KernelSnapshot extends Target {
       dartDefines: decodeDartDefines(environment.defines, kDartDefines),
       packageConfig: packageConfig,
+      buildDir: environment.buildDir,
+      generateDartPluginRegistry: environment.generateDartPluginRegistry,
```
The intention behind this flag is to enable the behavior in the environment where it makes sense. For instance, certain type of tests don't need to deal with the internals of the Dart plugin registry.

jonahwilliams replied:
I'm thinking more from the perspective of de-risking this change, such as using a feature or command line flag

blasten replied:
Sure. I can add a flutter config flag, which is enabled by default. wdyt?

jonahwilliams replied:
SGTM

zanderso replied:
I'm confused. Why does this have to be generated when it isn't needed at all? Don't we know enough in the tool to activate this only when there is some plugin present that wants it?

blasten replied:
> Why does this have to be generated when it isn't needed at all?

That is already the case. Please take a look at this line: https://github.com/flutter/flutter/blob/16c202341e9d5d5418cc7e2026ede9bcea4b80eb/packages/flutter_tools/lib/src/plugins.dart#L910

This is an additional mechanism to go a step further and disable the pubspec scan.

blasten commented on packages/flutter_tools/lib/src/dart/language_version.dart:
```diff
@@ -31,7 +40,7 @@ LanguageVersion determineLanguageVersion(File file, Package package) {
   try {
     lines = file.readAsLinesSync();
   } on FileSystemException {
-    return nullSafeVersion;
```
Could this be addressed in a separate PR? I mostly just added the notion to change the fallback version.

jonahwilliams replied:
No, this is a significant deviation in behavior from how language versions are supposed to behave

blasten replied:
The default value is still `nullSafeVersion`.  Are you referring to the generated main.dart?

jonahwilliams replied:
yes, why the change to fallback to optOutNullSafeVersion?

blasten replied:
I have an app that I haven't migrated to null safe. I flutter run it, and it complained that it cannot run in null safe mode.

jonahwilliams replied:
Then some part of the logic for filling in this language version is not working correctly - it should be able to pull the language version from your pubspec.yaml if there is no comment in the existing entrypoint.

blasten replied:
you right. I think I had a separate issue, so this change isn't required after all. Done.

Commit:
Review

Commit:
Fix test

Commit:
Revert changes to language version

Commit:
Mock dependency

Commit:
Merge branch 'master' into tool_dart_plugin_registrant

Commit:
Simplify a bit

jonahwilliams commented:
Shipping initial support for this and losing the ability to remove plugins from a pubspec doesn't seem like a reasonable tradeoff to me. I don't really have the full picture on how important this feature is, so I'd let @zanderso make a call as to whether that is reasonable

jonahwilliams commented:
The way that I would implement correct rebuilding of the registrant is to hoist the generation into a new `Target` instance, make the inputs the pubspec.yaml, the output the generated file. Then make `KernelSnapshot` depend on it.

You could re-use the same rule in the resident runner, with the localizations as an example of how to hook it up. Then all of the existing invalidation logic would work with it

zanderso commented on packages/flutter_tools/test/general.shard/plugins_test.dart:
```diff
@@ -1148,6 +1156,105 @@ flutter:
+            'import \'package:url_launcher_macos${fs.path.separator}url_launcher_macos.dart\';\n'
+            'import \'package:url_launcher_windows${fs.path.separator}url_launcher_windows.dart\';\n'
+            '\n'
+            '@pragma(\'vm:entry-point\')\n'
```
I don't understand why you need an additional entrypoint when you're wrapping main anyway. Couldn't `main()` below just be:
```
void main() {
  _registerPlugins();
  entrypoint.main();
}
```

blasten replied:
The embedding API in the engine allows to change the main entrypoint function. As a result, there's no way to know which entrypoint is invoked at runtime. It turns out that `main` is the default.

zanderso reviewed (changes requested):
Agree with Jonah we should be using the tool's build system to regenerate generated files whenever their deps change before trying to recompile.

stuartmorgan commented on packages/flutter_tools/test/general.shard/plugins_test.dart:
```diff
@@ -1148,6 +1156,105 @@ flutter:
+            ..createSync(recursive: true);
+
+        final Map<String, String> plugins = <String, String>{};
+        plugins['url_launcher_macos'] = '''
```
url_launcher isn't Dart-only; path_provider and shared_preferences would work for Linux and Windows.

We have no Dart-only macOS plugins.

blasten replied:
Just to clarify. This test is using a Memory file system, so it's not the real plugin you would normally install via pub get.  I pick these names for illustration purposes. I will add another Dart-only plugin to the test.

Commit:
Make test data more accurate

blasten commented on packages/flutter_tools/test/general.shard/plugins_test.dart:
```diff
@@ -1148,6 +1156,105 @@ flutter:
+            '\n'
+            '// @dart = 2.8\n'
+            '\n'
+            'import \'main\' as entrypoint;\n'
```
I updated the test data, so this looks more realistic to how it would work in the real environment.

Commit:
Add additional plugin to test

Commit:
Merge branch 'master' into tool_dart_plugin_registrant

blasten commented:
I'd really like to merge this PR this week. I initially sent a small patch 453c320 to get feedback,  and make everyone's review easier.  Please let me know how you'd like to proceed.

zanderso commented on packages/flutter_tools/lib/src/plugins.dart:
```diff
@@ -295,11 +322,21 @@ class Plugin {
+    if (!platformsYaml.containsKey(platformKey)) {
+      return false;
+    }
+    if ((platformsYaml[platformKey] as YamlMap).containsKey(kDefaultPackage)) {
```
Should this check that `platformsYaml[platformKey]` is a `YamlMap` first before doing the `as`?  Is that checked in a validate step somewhere else?

blasten replied:
Done!

zanderso reviewed (commented):
As long as this PR doesn't introduce a tool crash on a reload with an added/removed plugin, I'm okay with handling in a follow-up PR. Please file an issue and note it in a TODO(blasten).

One question about yaml validation below. I'd feel better if there was a test with invalid yaml for the new parsing to check that the validation and error handling is as expected.

I'll follow up with @jonahwilliams about my question about error handling in the build system. I want to make sure that this PR is getting everything out of it that it can because that will make the best user experience.

jonahwilliams commented on packages/flutter_tools/lib/src/plugins.dart:
```diff
@@ -360,14 +406,146 @@ Future<List<Plugin>> findPlugins(FlutterProject project, { bool throwOnError = t
+    }
+  }
+  if (didFindError && throwOnPluginPubspecError) {
+    throwToolExit('Please resolve the errors');
```
This should be handled automatically by the build system (see https://github.com/flutter/flutter/pull/75251 ) - the only tasks that get skipped during a failure are those that depend on failed tasks. sibling tasks aren't affected

blasten commented:
> As long as this PR doesn't introduce a tool crash on a reload with an added/removed plugin, I'm okay with handling in a follow-up PR. Please file an issue and note it in a TODO(blasten).

Adding or removing the plugin from pubspec.yaml doesn't trigger a crash as Hot restart doesn't fetch new dependencies. 

The following sequence causes an error, but no crash:
1. add plugin to pubspec.yaml
2. flutter run
3. remove plugin from pubspec.yaml
4. flutter pub get
5. Trigger hot restart

```
Error: Could not resolve the package 'url_launcher_linux' in 'package:url_launcher_linux/url_launcher_linux.dart'.
Error: Could not resolve the package 'url_launcher_macos' in 'package:url_launcher_macos/url_launcher_macos.dart'.
.dart_tool/flutter_build/generated_main.dart:9:8: Error: Not found: 'package:url_launcher_linux/url_launcher_linux.dart'
import 'package:url_launcher_linux/url_launcher_linux.dart';
       ^
.dart_tool/flutter_build/generated_main.dart:10:8: Error: Not found: 'package:url_launcher_macos/url_launcher_macos.dart'
import 'package:url_launcher_macos/url_launcher_macos.dart';
       ^
.dart_tool/flutter_build/generated_main.dart:15:7: Error: Getter not found: 'UrlLauncherPlugin'.
      UrlLauncherPlugin.registerWith();
      ^^^^^^^^^^^^^^^^^
.dart_tool/flutter_build/generated_main.dart:17:7: Error: Getter not found: 'UrlLauncherPlugin'.
      UrlLauncherPlugin.registerWith();
      ^^^^^^^^^^^^^^^^^
Performing hot restart...                                              ⣾Error: Could not resolve the package 'url_launcher_linux' in 'package:url_launcher_linux/url_launcher_linux.dart'.
Error: Could not resolve the package 'url_launcher_macos' in 'package:url_launcher_macos/url_launcher_macos.dart'.
.dart_tool/flutter_build/generated_main.dart:9:8: Error: Not found: 'package:url_launcher_linux/url_launcher_linux.dart'
import 'package:url_launcher_linux/url_launcher_linux.dart';
       ^
.dart_tool/flutter_build/generated_main.dart:10:8: Error: Not found: 'package:url_launcher_macos/url_launcher_macos.dart'
import 'package:url_launcher_macos/url_launcher_macos.dart';
       ^
.dart_tool/flutter_build/generated_main.dart:15:7: Error: Getter not found: 'UrlLauncherPlugin'.
      UrlLauncherPlugin.registerWith();
      ^^^^^^^^^^^^^^^^^
.dart_tool/flutter_build/generated_main.dart:17:7: Error: Getter not found: 'UrlLauncherPlugin'.
      UrlLauncherPlugin.registerWith();
      ^^^^^^^^^^^^^^^^^
```

zanderso commented:
@blasten Thanks for checking =) The only open items on my side are adding the TODO and the question about yaml validation.

jonahwilliams commented:
I don't really agree. The tool may not be crashing, but the user has no way of fixing that error without quitting and restarting - it might as well be a crash. If you think that change would be better done as a follow up because of the scope/size, I'm open to it, but we can't ship another stable without that implemented.

blasten commented:
> I don't really agree. The tool may not be crashing, but the user has no way of fixing that error without quitting and restarting - it might as well be a crash. 

I think this is currently true for a plugin that supports iOS and/or Android. For example, if you remove a plugin that provides implementation for a method channel,  then wouldn't that cause an exception if the platform implementation tries to reach the Dart code that no longer exists? 

Regenerating the main dart file on hot restart isn't hard, I implemented that originally. But I removed it after your comment https://github.com/flutter/flutter/pull/74469#discussion_r563975360 as doing it without changing those tests involve additional changes, and therefore expands the scope of this PR.

jonahwilliams commented:
From offline discussion, we can move forward with this PR as is, provided that the plugin removal issue is addressed as a release blocker

jonahwilliams commented on packages/flutter_tools/lib/src/plugins.dart:
```diff
@@ -684,6 +862,67 @@ Future<void> _writeAndroidPluginRegistrant(FlutterProject project, List<Plugin>
+  FlutterProject rootProject,
+  String currentMainUri,
+  File newMainDart,
+  File mainFile,
```
You can accept a PackageConfig instead of parsing it again, since the kernel compiler already has one

blasten replied:
Done!

jonahwilliams reviewed (approved):
LGTM if LG to @zanderso

zanderso commented:
From the PR review meeting: I think this one is still waiting on some of the items in https://github.com/flutter/flutter/pull/74469#pullrequestreview-581518355. Please correct me if I'm mistaken!

Commit:
Merge branch 'master' into tool_dart_plugin_registrant

Commit:
Test invalid pubpsec

zanderso reviewed (approved):
lgtm

jmagman commented:
Re: test failures

I added `url_launcher:` and `url_launcher_macos:` to a project pubspec.  Before this change, `.flutter-plugins-dependencies` looks like:`
```
  "plugins": {
    "ios": [
      {
        "name": "url_launcher",
        "path": "{path}",
        "dependencies": []
      }
    ],
    "android": [
      {
        "name": "url_launcher",
        "path": "{path}",
        "dependencies": []
      }
    ],
    "macos": [
      {
        "name": "url_launcher_macos",
        "path": "{path}",
        "dependencies": []
      }
    ],
    "linux": [
      {
        "name": "url_launcher_linux",
        "path": "{path}",
        "dependencies": []
      }
    ],
    "windows": [
      {
        "name": "url_launcher_windows",
        "path": "{path}",
        "dependencies": []
      }
    ],
    "web": [
      {
        "name": "url_launcher_web",
        "path": "{path}",
        "dependencies": []
      }
    ]
  },
```
On this PR `url_launcher` is listed as a plugin for every platform, with the platform-specific variant as a dependency.
```
  "plugins": {
    "ios": [
      {
        "name": "url_launcher",
        "path": "{path}",
        "dependencies": []
      }
    ],
    "android": [
      {
        "name": "url_launcher",
        "path": "{path}",
        "dependencies": []
      }
    ],
    "macos": [
      {
        "name": "url_launcher",
        "path": "{path}",
        "dependencies": [
          "url_launcher_macos"
        ]
      },
      {
        "name": "url_launcher_macos",
        "path": "{path}",
        "dependencies": []
      }
    ],
    "linux": [
      {
        "name": "url_launcher",
        "path": "{path}",
        "dependencies": [
          "url_launcher_linux"
        ]
      },
      {
        "name": "url_launcher_linux",
        "path": "{path}",
        "dependencies": []
      }
    ],
    "windows": [
      {
        "name": "url_launcher",
        "path": "{path}",
        "dependencies": [
          "url_launcher_windows"
        ]
      },
      {
        "name": "url_launcher_windows",
        "path": "{path}",
        "dependencies": []
      }
    ],
    "web": [
      {
        "name": "url_launcher_web",
        "path": "{path}",
        "dependencies": []
      }
    ]
  },
```
If that's the intended behavior of this PR, those tests should be updated.

blasten commented:
Thanks @jmagman. That seems correct based on my changes. (which I had to review again :))

Commit:
Update test

Commit:
Merge branch 'master' into tool_dart_plugin_registrant

jmagman commented on dev/devicelab/bin/tasks/plugin_lint_mac.dart:
```diff
@@ -473,7 +473,7 @@ void _validateMacOSPodfile(String appPath) {
       || !podfileLockOutput.contains(':path: Flutter/ephemeral/.symlinks/plugins/url_launcher_macos/macos')
       || !podfileLockOutput.contains(':path: Flutter/ephemeral/.symlinks/plugins/test_plugin_swift/macos')
-      || podfileLockOutput.contains('url_launcher/')) {
+      || !podfileLockOutput.contains('url_launcher/')) {
```
The spirit of this check was to make sure the `url_launcher` framework was not embedded in the app.  It was using the Podfile.lock as a shortcut for this.  A more appropriate test now would be to validate that the generated app contains `Contents/Frameworks/url_launcher_macos.framework` but not `Contents/Frameworks/url_launcher.framework`.

blasten replied:
got it. I added another check to the test for the time being. Do you have any suggestion about how to verify the frameworks?

jmagman replied:
Something like
```dart
  final String appDirectory = path.join(
    appPath,
    'build',
    'macos',
    'Build',
    'Products',
    'Release',
    '${path.basename(appPath)}.app',
  );

  final String frameworksDirectory = path.join(
    appDirectory,
    'Contents',
    'Frameworks',
  );

  checkDirectoryExists(path.join(
    frameworksDirectory,
    'url_launcher_macos.framework',
  ));

  checkDirectoryNotExists(path.join(
    frameworksDirectory,
    'url_launcher.framework',
  ));
```

There also may be a more appropriate test location for that check, like https://github.com/flutter/flutter/blob/master/packages/flutter_tools/test/integration.shard/macos_content_validation_test.dart

Commit:
Check specific import

Commit:
Merge branch 'master' into tool_dart_plugin_registrant

Commit:
Rework what it means to support a platform as current approach caused infinite recursion in analyzing packages

Commit:
Revert changes to plugin_lint_mac

Commit:
Update tests
